src/ddscxx/src/org/eclipse/cyclonedds/core/cond/ConditionDelegate.cpp

```diff
@@ -39,17 +39,47 @@ org::eclipse::cyclonedds::core::cond::ConditionDelegate::~ConditionDelegate()
     // sure that functor is delete here
     if (this->myFunctor) {
         delete this->myFunctor;
+        this->myFunctor = nullptr;
     }
 }
 
+void
+org::eclipse::cyclonedds::core::cond::ConditionDelegate::detach_from_waitset(
+    const dds_entity_t entity_handle) {
+    std::vector<WaitSetDelegate *> waitset_list_tmp;
+    org::eclipse::cyclonedds::core::ScopedMutexLock scopedLockForCopy(this->waitSetListUpdateMutex);
+    waitset_list_tmp.assign(this->waitSetList.begin(), this->waitSetList.end());
+    scopedLockForCopy.unlock();
+
+    for (auto waitset : waitset_list_tmp) {
+        org::eclipse::cyclonedds::core::ScopedObjectLock scopedWaisetLock(*waitset);
+        {
+            org::eclipse::cyclonedds::core::ScopedMutexLock scopedLock(this->waitSetListUpdateMutex);
+            // remove the waitset from the list and detach the condition
+            if (this->waitSetList.erase(waitset)) {
+                waitset->remove_condition_locked(this, entity_handle);
+            }
+        }
+    }
+}
+
+void
+org::eclipse::cyclonedds::core::cond::ConditionDelegate::detach_and_close(
+    const dds_entity_t entity_handle)
+{
+    detach_from_waitset(entity_handle);
+    org::eclipse::cyclonedds::core::cond::ConditionDelegate::close();
+}
+
 void
 org::eclipse::cyclonedds::core::cond::ConditionDelegate::close()
```

Comment:
The reason for adding a new API `detach_and_close()` is to get the entity handle without calling `get_ddsc_entity()` as it will lead to a double lock, and can't modify the `ConditionDelegate::close()` with some default parameter as it will hide the virtual interface function.

Reply:
Perfectly sensible. I suspect there is still a window for a race condition where one can sneak in an `attach` in between copying the list of waitsets and looping over them. Given the set of problems this PR addresses and the obvious API abuse involved, I can't be too fussed about not closing that tiny hole in this PR.